### PR TITLE
Add support to cache "all" profiles

### PR DESCRIPTION
**Pull request by shivamerla**

### CODE REVIEW

**Comment by mikemckiernan** [comment]
This is going to be fun.

```
69G     nim-service-model-store-pvc-6c048988-2371-4a2b-ae30-107f31560ab5/
```

It's still going.

**Comment by shivamerla** [comment]
> This is going to be fun.
> 
> ```
> 69G     nim-service-model-store-pvc-6c048988-2371-4a2b-ae30-107f31560ab5/
> ```
> 
> It's still going.

Give it a day it will be done 😜

**Comment by mikemckiernan** [comment]
I ran out of disk at 169G.

> Give it a day it will be done 😜

File: internal/controller/nimcache_controller.go [hunk 1]
```diff
@@ -463,7 +464,7 @@ func (r *NIMCacheReconciler) reconcileJobStatus(ctx context.Context, nimCache *a
 			return fmt.Errorf("failed to get selected profiles: %w", err)
 		}
 
-		if len(selectedProfiles) > 0 {
+		if len(selectedProfiles) > 0 && !utils.ContainsElement(selectedProfiles, "all") {
```
Comment (slu2011) [inline internal/controller/nimcache_controller.go:467]: Add a const string definition for "all"?

Comment (elezar) [inline internal/controller/nimcache_controller.go:467]: Does `selectedProfiles` represent a concrete type. Does it make sense to add a function like `IsAll()` to capture this logic?

File: internal/controller/nimcache_controller.go [hunk 2]
```diff
@@ -816,9 +817,14 @@ func constructJob(nimCache *appsv1alpha1.NIMCache) (*batchv1.Job, error) {
...
-		if selectedProfiles != nil {
-			job.Spec.Template.Spec.Containers[0].Args = []string{"--profiles"}
-			job.Spec.Template.Spec.Containers[0].Args = append(job.Spec.Template.Spec.Containers[0].Args, selectedProfiles...)
+
+		if len(selectedProfiles) > 0 {
+			if utils.ContainsElement(selectedProfiles, "all") {
+				job.Spec.Template.Spec.Containers[0].Args = []string{"--all"}
+			} else {
+				job.Spec.Template.Spec.Containers[0].Args = []string{"--profiles"}
+				job.Spec.Template.Spec.Containers[0].Args = append(job.Spec.Template.Spec.Containers[0].Args, selectedProfiles...)
```
Comment (elezar) [inline internal/controller/nimcache_controller.go:826]: should the profiles be comma-separated?

File: internal/controller/nimcache_controller_test.go [hunk 3]
```diff
@@ -282,6 +282,28 @@ var _ = Describe("NIMCache Controller", func() {
...
+			}
+
+			job, err := constructJob(nimCache)
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(job.Name).To(Equal(getJobName(nimCache)))
+			Expect(job.Spec.Template.Spec.Containers[0].Image).To(Equal("nvcr.io/nim:test"))
+			Expect(job.Spec.Template.Spec.ImagePullSecrets[0].Name).To(Equal("my-secret"))
+			Expect(job.Spec.Template.Spec.Containers[0].Command).To(ContainElements("download-to-cache"))
+			Expect(job.Spec.Template.Spec.Containers[0].Args).To(ContainElements("--all"))
```
Comment (elezar) [inline internal/controller/nimcache_controller_test.go:304]: Do we have a test for a specific profile as well as multiple profiles?

File: internal/controller/nimcache_controller.go (hunk 1 above)
Comment (shivamerla) [reply internal/controller/nimcache_controller.go:467]: yes, it is a concrete type, but adding a utility call for this makes sense.

File: internal/controller/nimcache_controller.go (hunk 2 above)
Comment (shivamerla) [reply internal/controller/nimcache_controller.go:826]: no, they are space delimited currently by the way how NIM container is parsing the CLI args. All arguments following `--profiles` are considered as profiles. We would need to update to comma separated if that logic changes.

File: internal/controller/nimcache_controller_test.go (hunk 3 above)
Comment (shivamerla) [reply internal/controller/nimcache_controller_test.go:304]: not currently, could easily add that.

File: internal/controller/nimcache_controller.go (hunk 1 above)
Comment (shivamerla) [reply internal/controller/nimcache_controller.go:467]: sure

Comment (shivamerla) [reply internal/controller/nimcache_controller.go]: double checked again and this is confined to NIMCache type as "all" profiles is not used during serving. Keeping the code as-is.

**Review by ArangoGutierrez — APPROVED** [review]
LGTM

**Review by slu2011 — APPROVED** [review]

**Pull request merged**